FWPositionController: improve estimation of true airspeed derivative

sfuhrer opened the pull request:
### Solved Problem
Bad airspeed tracking when flying with high angle of attack or when slipping in turns, most noticeable in high bank turns. Cause is a bad airspeed rate estimation when the vehicle attitude is perfectly aligned with direction of travel, as currently the groundspeed derivative in vehicle x-direction is taken into account.
![image](https://user-images.githubusercontent.com/26798987/223778106-85ff453f-7869-43c6-8c4b-1fbc32a5f1a3.png)

### Solution
Instead of using acceleration in body x direction, use acceleration in direction of flight for true airspeed derivative estimate.

Here a SITL flight with increasingly high roll angle. See in second plot the comparison between old and new way of estimating TAS_rate.
![image](https://user-images.githubusercontent.com/26798987/223777947-95d652b8-b301-49aa-b6d2-4c130a5c1b29.png)

a7977ccd60e0e9b9e64ee74f2b640089e8d17880 is kind of unrelated, I feel like we don't need to store the whole vectors globally.

sfuhrer committed:
FWPositionController: only store acceleration and velocity in x

Signed-off-by: Silvan Fuhrer <silvan@auterion.com>

sfuhrer committed:
FWPositionController: change way of estimating airspeed derivative

Use local_position.a in direction of local_position.v - wind when wind is valid,
and otherwise simply input 0 TAS_rate_estimate to TECS.

Signed-off-by: Silvan Fuhrer <silvan@auterion.com>

tstastny commented on src/modules/fw_path_navigation/FixedwingPositionControl.cpp:
```diff
@@ -2550,6 +2553,10 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 
+	const Vector3f local_pos_a = Vector3f{_local_pos.ax, _local_pos.ay, _local_pos.az};
+	const Vector3f local_pos_v = Vector3f{_local_pos.vx, _local_pos.vy, _local_pos.vz};
+	const float acceleration_in_direction_of_flight = local_pos_v.unit_or_zero().dot(local_pos_a);
```
Why into direction of flight? don't we want in direction of airspeed vector? - so subtracting the wind estimate off the ground vel vector.

RomanBapst replied:
@tstastny @sfuhrer That would at least in theory give you the correct "airspeed derivative". When you fly a circle with wind then the vehicle is constantly accelerating / decelerating  depending on where it is on the circle. Measuring the inertial acceleration into the wind direction should more or less cancel out the acceleration as the relative wind vector is more or less constant in magnitude, right?

RomanBapst replied:
@sfuhrer Have you tried this solution in SITL with wind as well?

sfuhrer replied:
Yes I probably neglected wind a bit too much.
Anyway, I'm currently looking into removing the acceleration input into TECS completely, as before the fix https://github.com/PX4/PX4-Autopilot/pull/20443/commits/b1cbe04694bcb85161126df641dc7ebb66e8697a got in that was roughly how TECS functioned (the internal TAS_rate state was always close to 0). And it flew just fine, actually better then it does now.

tstastny replied:
> Anyway, I'm currently looking into removing the acceleration input into TECS completely, as before the fix https://github.com/PX4/PX4-Autopilot/commit/b1cbe04694bcb85161126df641dc7ebb66e8697a got in that was roughly how TECS functioned (the internal TAS_rate state was always close to 0). And it flew just fine, actually better then it does now.

Should we just zero the accel input for now then, e.g. for 1.14 release, such that there is no change in behavior / regression - then we re-include accel once we've had time to make sure it is fedback correctly and tested?

sfuhrer committed:
TECS: remove ununsed getSteRate()

Signed-off-by: Silvan Fuhrer <silvan@auterion.com>

sfuhrer commented:
I did some performance comparison for tracking a constant airspeed during a increasingly tight loiter (80m radius down to 30m).

Flight 1: current TECS
![image](https://user-images.githubusercontent.com/26798987/224953269-417845d8-1139-4ef2-a268-59c7dd8c29f8.png)


Flight 2: set the measured TAS derivative (as input into TECS) to 0
![image](https://user-images.githubusercontent.com/26798987/224954166-acd307b9-2b0b-4c7f-b36c-bba534a93b73.png)
![image](https://user-images.githubusercontent.com/26798987/224958248-2a0aacff-df7c-4d44-8db1-6d0ec6093f1f.png)

Flight 3: set TAS derivative to `groundspeed_minus_wind.unit_or_zero().dot(local_pos_a)`
![image](https://user-images.githubusercontent.com/26798987/224955496-ea51c9a0-7542-4f46-8e4b-9f40b349c4d6.png)
![image](https://user-images.githubusercontent.com/26798987/224958070-4fa3ed2a-e4c7-4f4b-a6b1-4d44ac07f708.png)


Conclusion: It's very obvious that current TECS with the acceleration simply taken in direction of heading instead of direction of airspeed vector has very poor airspeed tracking in high bank. 
Simply discarding the acceleration input into TECS (setting it to 0) improves the situation drastically, it now tracks airspeeds without offset over the whole bank range. One can though see the higher delay in tecs_status/true_airspeed_filtered compared to to the airspeed sensor data, and thus also a higher delay in airspeed setpoint tracking. 
Flight 3, with the airspeed derivative estimated as the dot product of groundspeed (minus wind) and acceleration, yields the best results, the airspeed tracking is very tight.

The next thing I checked was resilience of the here proposed airspeed rate estimate source to GNSS failures. For that I disabled GPS fusion (EKF2_GPS_CTRL to 0) in flight (the vehicle then stayed estimating it's position through wind dead-reckoning). 
The following plots show the outcome. The increasing` local_position/eph `indicates when GNSS fusion was switched off (for 3.5minutes). No deteriorating airspeed tracking is noticeable, thus indicating that` local_position.velocity` is well estimated also without GNSS fusion. 
![image](https://user-images.githubusercontent.com/26798987/224961366-8d562ef0-1191-4773-b0ee-1e9a25c11505.png)

tstastny commented:
> The increasinglocal_position/ephindicates when GNSS fusion was switched off (for 3.5minutes). No deteriorating airspeed tracking is noticeable, thus indicating that local_position.velocity is well estimated also without GNSS fusion.

so the only last point is to determine if there is any classifiable degradation of the local velocity signal during dead reckoning that would merit additionally switching off the acceleration feedback to TECS, right?

RomanBapst commented:
> so the only last point is to determine if there is any classifiable degradation of the local velocity signal during dead reckoning that would merit additionally switching off the acceleration feedback to TECS, right?

@sfuhrer @tstastny Also we should monitor and be aware that the ground minus wind direction approach could have large initial error for a small period of time when wind states have not yet converged. So wet could potentially also require a condition for wind states convergence.

sfuhrer commented:
>so the only last point is to determine if there is any classifiable degradation of the local velocity signal during dead reckoning that would merit additionally switching off the acceleration feedback to TECS, right?

As discussed offline - we so far have seen no indication that the groundspeed vector through local_position.v gets bad while the system still is declaring position estimation valid, so we for now don't want to add an extra check on velocity degradation. 

>Also we should monitor and be aware that the ground minus wind direction approach could have large initial error for a small period of time when wind states have not yet converged. So wet could potentially also require a condition for wind states convergence.

Fair point. I currently use the _wind_valid state for using wind or not, which in turn is simply validated if there is a finite wind estimate received and invalidated with a 10s timeout. 
We should then thus work on improving the checks for _wind_valid. One option is to add a similar threshold for the variance as for using the wind speed for determine airspeed in the airspeed selector, https://github.com/PX4/PX4-Autopilot/blob/4ff262abde7ceb44882007615a4f914349ce6c5e/src/modules/airspeed_selector/airspeed_selector_main.cpp#L550.
One other idea would be to assume that wind is only off during initialization, and a simple work around for that would be to only validate _wind_valid some 10s or so after the initialization.

tstastny commented:
> and a simple work around for that would be to only validate _wind_valid some 10s or so after the initialization.

would really try to avoid arbitrary second thresholds. much prefer if we can define a physical relationship here -- uncertainty has physical units. only maybe small time hysteresis term if necessary.. eg if @RomanBapst @sfuhrer you have seen and expect the uncertainty value to oscillate around whatever threshold ever.

sfuhrer committed:
FWPosControl: Only use accleration input for TECS if local_position has valid velocity

And additionally only if the control mode flag position is set, as then we can assume
that all the fields in local_position are valid.

Signed-off-by: Silvan Fuhrer <silvan@auterion.com>

dagar commented on src/modules/fw_path_navigation/FixedwingPositionControl.cpp:
```diff
@@ -2550,6 +2553,13 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 					 throttle_max);
 
+	const Vector3f local_pos_a = Vector3f{_local_pos.ax, _local_pos.ay, _local_pos.az};
+	const Vector3f groundspee_minus_wind = Vector3f{_local_pos.vx - _wind_vel(0), _local_pos.vy - _wind_vel(1), _local_pos.vz};
```
```suggestion
	const Vector3f groundspeed_minus_wind = Vector3f{_local_pos.vx - _wind_vel(0), _local_pos.vy - _wind_vel(1), _local_pos.vz};
```

sfuhrer commented:
I checked some logs for when I would start using a wind estimate (indicated with cursor). A variance threshold of 0.05 seems to be reasonable. 
![image](https://user-images.githubusercontent.com/26798987/225058020-22e6dacc-e6f2-41ba-9cfc-8bd9ad3c17b5.png)
![image](https://user-images.githubusercontent.com/26798987/225058476-2bd774d3-f61d-4826-b310-a445a253d5e7.png)
![image](https://user-images.githubusercontent.com/26798987/225058901-12cd9580-1f91-43f5-b81c-21a925d0a610.png)
![image](https://user-images.githubusercontent.com/26798987/225059357-36534545-4d59-4872-8bf9-0da08635a68f.png)
![image](https://user-images.githubusercontent.com/26798987/225061020-2516e8c9-e551-4b40-887d-3e1f98f8795f.png)
![image](https://user-images.githubusercontent.com/26798987/225061631-47adcb08-ba29-4d76-9480-86d5341c8d6d.png)
![image](https://user-images.githubusercontent.com/26798987/225067649-0715a4dc-a568-44f7-837b-772894f00a42.png)
![image](https://user-images.githubusercontent.com/26798987/225068181-56df1c1a-71a7-4773-82c2-78c22a4758d6.png)

sfuhrer committed:
FWPositionControl: add extra check for low wind variance to use TAS rate estimate in TECS

Signed-off-by: Silvan Fuhrer <silvan@auterion.com>

sfuhrer commented:
I think the wind validity checks need more thought and testing, plus possibly a bit of architectural changes to facilitate it cleanly. I thus propose we hotfix the issue with the alternative I proposed above (0 the airspeed rate input). See https://github.com/PX4/PX4-Autopilot/pull/21317
